boards: Add support for the NXP MIMXRT595 DSP core

dluke62 opened the pull request:
Add board and soc files for the audio DSP in NXP MIMXRT595.

DerekSnell reviewed (changes requested):
Hi @dluke62 ,
Thank you for contributing this!  Can you fix the compliance issues?  I also have some quick feedback of the SOC series naming convention we need to change.  I'm sure we will have more feedback to come.  Thank you

dbaluta commented on soc/xtensa/nxp_adsp/Kconfig:
```diff
@@ -14,6 +14,10 @@ config SOC_FAMILY
+	default y
+
 # Select SoC Part No. and configuration options
 source "soc/xtensa/nxp_adsp/*/Kconfig.soc"
```
@dluke62 I wonder how big of an effort would be for you to split this PR into multiple patches. 

For example, I would like to see the introduction of NXP_ADSP_SIGN_IMAGE and preparation to accomodate RT in a separate patch.

Usually we do cleanups and preparation patches first and then add a big patch with the support for a new boards. 

What do you think?

dluke62 replied:
hi @dbaluta , thank you for your comment, I think i can split the PR in two commits - first one will update the Kconfigs in preparation for RT SoC, second one will add the board/soc files for RT

dbaluta replied:
@dluke62 thanks. That seems reasonable. Usually, we add one patch per logical change.

dbaluta commented on soc/xtensa/nxp_adsp/imx8m/Kconfig.defconfig.series:
```diff
@@ -11,44 +11,20 @@ config SOC_TOOLCHAIN_NAME
 
-config SMP
-	default n
-
```
Same here, I think this kind of changes should be placed in a separate patch with an explanation so that later when we go and try to understand the changes in git history everything will be simple and clear.

dluke62 replied:
Updated per your recommendation. Thank you!

dbaluta commented:
Also, I guess you will need to add at least some compile time tests. @iuliana-prodan care to provide some guidance here on how did you do this for imx8m_adsp?

iuliana-prodan commented:
> Also, I guess you will need to add at least some compile time tests. @iuliana-prodan care to provide some guidance here on how did you do this for imx8m_adsp?

You can have a simple compilation test by adding a yaml file in the `boards` folder for your target: see https://docs.zephyrproject.org/latest/develop/test/twister.html and an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m.yaml

dluke62 commented on soc/xtensa/nxp_adsp/imx8m/Kconfig.defconfig.series:
```diff
@@ -11,44 +11,20 @@ config SOC_TOOLCHAIN_NAME
 	string
 	default "nxp_imx8m_adsp"
 
-# if SOC_MIMX8M_ADSP
```
it is commented out, it might have been copy/paste from another Kconfig.

iuliana-prodan replied:
no, is for this soc, for sure. 
It has nothing to do with the rest of the changes, that's why I don't understand why you removed it.
If you want, add a clean-up/refactor patch separately.

dbaluta replied:
@dluke62 @iuliana-prodan Let's this do the proper way.

So, patch 2 with the refactor and preparing kconfig files for RT adsp should be split into the following patches:

1) Introduce CONFIG_NXP_ADSP_SIGN_IMAGE that allows you to sign the image.
2) Move common configs into one common place.
3) Remove configs symbols that are already on the correct default value.

Each patch should have proper explanation stressing on WHY the patch is needed.

dluke62 replied:
Updated commits leaving 2. out. This can be a separate cleanup pull request. No need to do it here.

dluke62 commented on soc/xtensa/nxp_adsp/Kconfig:
```diff
@@ -14,6 +14,10 @@ config SOC_FAMILY
 	string
 	default "nxp_adsp"
 
+config NXP_ADSP_SIGN_IMAGE
```
hi @iuliana-prodan , if you don't sign NXP images, can we simply remove the custom command from nxp_adsp/CMakeLists.txt and not add the NXP_ADSP_SIGN_IMAGE config?

iuliana-prodan replied:
Hi @dluke62,

I know it seems reasonable to remove it if not used, but we added to be in sync with Intel, since the DSP (at least for nxp_adsp_imx8m and nxp_adsp_imx8) is used for SOF (https://thesofproject.github.io/latest/architectures/firmware/sof-zephyr/index.html). We are not signing the images now, but we might.

Please, let's keep it, for now. 
If we remove it, we need more testing for all platforms.

Also, just FYI, for nxp_adsp_imx8m and  nxp_adsp_imx8 we are using rimage to create the SOF firmware (from the zephyr elf we add some custom headers and we obtain a .ri image).

dluke62 commented:
@iuliana-prodan , @dbaluta if I understand correctly, we can't have any compile time tests because we can't compile for RT595 DSP without downloading Xtensa toolchain separately (unlike imx8, which has toolchain in Zephyr SDK). Is that correct?

Commit:
boards: Add support for the NXP MIMXRT595 DSP core

Add board and soc files for the NXP MIMXRT595 DSP core.

Signed-off-by: Dmitry Lukyantsev <dmitrylu@google.com>

dleach02 commented:
@stephanosio how would one submit an enablement to Zephyr upstream if there isn't an open toolchain in our CI?

dbaluta commented:
> @iuliana-prodan , @dbaluta if I understand correctly, we can't have any compile time tests because we can't compile for RT595 DSP without downloading Xtensa toolchain separately (unlike imx8, which has toolchain in Zephyr SDK). So the CI won't be able to compile Zephyr for RT595 DSP. Is that correct?

Technically, once you have the Xtensa toolchain configuration you can use that to create the corresponding toolchain and integrate it in Zephyr SDK.

For example, here is how imx8 toolchain was added:

https://github.com/zephyrproject-rtos/sdk-ng/pull/187/commits

Another clean PR for adding a toolchain for a new board is here: 

https://github.com/zephyrproject-rtos/sdk-ng/pull/694

dluke62 commented:
> > @iuliana-prodan , @dbaluta if I understand correctly, we can't have any compile time tests because we can't compile for RT595 DSP without downloading Xtensa toolchain separately (unlike imx8, which has toolchain in Zephyr SDK). So the CI won't be able to compile Zephyr for RT595 DSP. Is that correct?
> 
> Technically, once you have the Xtensa toolchain configuration you can use that to create the corresponding toolchain and integrate it in Zephyr SDK.
> 
> For example, here is how imx8 toolchain was added:
> 
> https://github.com/zephyrproject-rtos/sdk-ng/pull/187/commits
> 
> Another clean PR for adding a toolchain for a new board is here:
> 
> [zephyrproject-rtos/sdk-ng#694](https://github.com/zephyrproject-rtos/sdk-ng/pull/694)

Thank you! As far as I understand the toolchain/core configuration is not freely available - one has to create an account with Cadence to get it. I don't know if we can simply check it in here - I'm assuming it comes with its own license, user agreement, etc. @DerekSnell could you please chime in?

iuliana-prodan commented on soc/xtensa/nxp_adsp/imx8/Kconfig.defconfig.series:
```diff
@@ -15,36 +15,16 @@ config SOC
 config SYS_CLOCK_TICKS_PER_SEC
 	default 50000
 
-config KERNEL_ENTRY
```
But if MULTI_LEVEL_INTERRUPTS depends on GEN_SW_ISR_TABLE, which depends on GEN_ISR_TABLES, which is set to Y, doesn't mean that MULTI_LEVEL_INTERRUPTS is Y?

dluke62 replied:
Thank you! I reverted the change and will keep these configs in common Kconfig.defconfig.

DerekSnell commented on soc/xtensa/nxp_adsp/rt595/Kconfig.soc:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2023 Google LLC.
+# SPDX-License-Identifier: Apache-2.0
+
+choice
+	prompt "NXP RT595 ADSP SoC Selection"
```
```suggestion
choice
	prompt "NXP RT5xx ADSP SoC Selection"
```

DerekSnell commented on soc/xtensa/nxp_adsp/rt595/Kconfig.soc:
```diff
@@ -0,0 +1,10 @@
+
+	config SOC_NXP_RT595
+		bool "NXP RT595"
+		depends on SOC_SERIES_NXP_RT595
```
```suggestion
		depends on SOC_SERIES_NXP_RT5XX
```

DerekSnell commented on boards/xtensa/nxp_adsp_rt595/Kconfig.board:
```diff
@@ -0,0 +1,6 @@
+
+config BOARD_NXP_ADSP_RT595
+	bool "NXP ADSP RT595"
+	depends on SOC_SERIES_NXP_RT595
```
```suggestion
	depends on SOC_SERIES_NXP_RT5XX
```

DerekSnell commented on boards/xtensa/nxp_adsp_rt595/nxp_adsp_rt595_defconfig:
```diff
@@ -0,0 +1,7 @@
+CONFIG_SOC_SERIES_NXP_RT595=y
```
```suggestion
CONFIG_SOC_SERIES_NXP_RT5XX=y
```

DerekSnell commented on soc/xtensa/nxp_adsp/CMakeLists.txt:
```diff
@@ -1,20 +1,26 @@
 # Copyright (c) 2021 NXP
 # SPDX-License-Identifier: Apache-2.0
 
+if(CONFIG_SOC_SERIES_NXP_RT595)
```
```suggestion
if(CONFIG_SOC_NXP_RT595)
```
Since these macro is for the SOC part number, this Kconfig should be the SOC name, not the SOC series.

DerekSnell commented on soc/xtensa/nxp_adsp/rt595/Kconfig.series:
```diff
@@ -0,0 +1,12 @@
+	select XTENSA_RESET_VECTOR
+	select XTENSA_USE_CORE_CRT1
+	help
+	  NXP RT595 ADSP Series
```
```suggestion
	  NXP RT5xx ADSP Series
```

DerekSnell reviewed (dismissed):
Hi @dluke62 ,
Sorry, yesterday I requested these changes when I submitted my review before you pushed an update.  But it appears GitHub did not capture them (probably my fault).

We need to change the naming of the SOC series.  RT595 is the SOC derivative in the i.MX RT500 series, which in Zephyr is named RT5xx.  NXP may offer other RT5xx part numbers in this series using the same F1 DSP core, which would share this code.  So my changes keep the RT595 name when referencing the SOC, but use RT5xx when referencing the series.

Thanks again for your contribution!

Commit:
soc: nxp_adsp: Refactor imx8/imx8m Kconfig

In preparation for RT500 ADSP enablement, consolidate common Xtensa
configuration parameters in top level Kconfig.defconfig.

Signed-off-by: Dmitry Lukyantsev <dmitrylu@google.com>

dluke62 commented on boards/arm/mimxrt595_evk/doc/index.rst:
```diff
@@ -174,6 +174,30 @@ Serial Port
+this. For instructions on how to download the toolchain and debug DSP firmware,
+please refer to this document: `Getting Started with Xplorer for EVK-MIMXRT595`_
+
+The following environment variables must be set prior to building RT595 DSP
```
fixed

dluke62 commented:
> @dbaluta, @DerekSnell  thank you for your comments, I've update the commits, I believe I addressed all the issues. Please take a look when you get a chance, let me know if there is anything else I should fix.
Since you guys are at NXP and this is an NXP part, would you be able to obtain the config file internally? Most likely I won't be able to send/share any files.
To your question how I load the firmware - I'm using Xtensa Xplorer + xt-ocd combo to load elf file via JTAG (while M33 core is running custom firmware that powers up the DSP core).

DerekSnell commented:
Hi @dluke62 ,
Thank you for addressing all the changes I requested.  @dbaluta is looking into adding the toolchain for this DSP core.  I will be out of office next week, and I do not want my block to hold this PR.  I am not approving yet since the toolchain is needed for CI to pass, and for us to test samples can build and run on this board.  But I will dismiss my review since you addressed all my other requests.  

Thanks again for all your work on this.

dbaluta commented:
@dluke62 thanks, will have a look and create the open toolchain within Zephyr SDK.

I will use newest configuration nxp_rt500_RI2021_8_newlib_linux.tgz. Hope this is OK with you.

dbaluta commented:
First try for the toolchain is here: https://github.com/zephyrproject-rtos/sdk-ng/pull/696

dluke62 commented:
> Hi @dluke62 , Thank you for addressing all the changes I requested. @dbaluta is looking into adding the toolchain for this DSP core. I will be out of office next week, and I do not want my block to hold this PR. I am not approving yet since the toolchain is needed for CI to pass, and for us to test samples can build and run on this board. But I will dismiss my review since you addressed all my other requests.
> 
> Thanks again for all your work on this.

Great, thank you very much @DerekSnell

dluke62 commented:
> @dluke62 thanks, will have a look and create the open toolchain within Zephyr SDK.
> 
> I will use newest configuration nxp_rt500_RI2021_8_newlib_linux.tgz. Hope this is OK with you.

Great, thank you very much for looking into this!

dleach02 commented:
> > @dluke62 thanks, will have a look and create the open toolchain within Zephyr SDK.
> > I will use newest configuration nxp_rt500_RI2021_8_newlib_linux.tgz. Hope this is OK with you.
> 
> Great, thank you very much for looking into this!

@dluke62 can you

marc-hb reviewed (dismissed):
About toolchains, sharing these hard to find links just in case they may help. Sorry if not
- https://docs.zephyrproject.org/latest/boards/xtensa/intel_adsp_cavs25/doc/intel_adsp_generic.html#xtensa-toolchain-optional
- https://thesofproject.github.io/latest/introduction/index.html#toolchain-faq
- https://thesofproject.github.io/latest/getting_started/build-guide/build-from-scratch.html#step-3-build-toolchains-from-source

dluke62 commented:
hi @marc-hb, thank you for your comments. I am getting this build error if I don't disable rimage step in CMakeLists.txt:

FAILED: zephyr/zephyr.ri zephyr/build/zephyr/zephyr.ri 
cd zephyr/build/zephyr/soc/xtensa/nxp_adsp && west sign --if-tool-available --tool rimage --build-dir zephyr/build
FATAL ERROR: rimage target not defined
ninja: build stopped: subcommand failed.

I don't have rimage on my machine, it's not in PATH. Build for imx8m succeeds, I'm getting the warning that rimage is not found.

iuliana-prodan commented:
> hi @marc-hb, thank you for your comments. I am getting this build error if I don't disable rimage step in CMakeLists.txt:
> 
> FAILED: zephyr/zephyr.ri zephyr/build/zephyr/zephyr.ri cd zephyr/build/zephyr/soc/xtensa/nxp_adsp && west sign --if-tool-available --tool rimage --build-dir zephyr/build FATAL ERROR: rimage target not defined ninja: build stopped: subcommand failed.
> 
> I don't have rimage on my machine, it's not in PATH. Build for imx8m succeeds, I'm getting the warning that rimage is not found.

Yes, that's because of this: https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/extensions.cmake#L832C10-L832C34, which is defined in each `board.cmake`:
```
~/data/zephyrproject/zephyr$ grep -rnw board_set_rimage_target 
cmake/modules/extensions.cmake:832:function(board_set_rimage_target target)
boards/xtensa/nxp_adsp_imx8x/board.cmake:6:board_set_rimage_target(imx8)
boards/xtensa/intel_adsp_ace20_lnl/board.cmake:5:board_set_rimage_target(lnl)
boards/xtensa/intel_adsp_ace15_mtpm/board.cmake:5:board_set_rimage_target(mtl)
boards/xtensa/intel_adsp_cavs25/board.cmake:13:board_set_rimage_target(tgl)
boards/xtensa/intel_adsp_cavs25/board.cmake:17:board_set_rimage_target(tgl-h)
boards/xtensa/nxp_adsp_imx8m/board.cmake:6:board_set_rimage_target(imx8m)
boards/xtensa/nxp_adsp_imx8/board.cmake:6:board_set_rimage_target(imx8)
```

Check this fix:
```
diff --git a/scripts/west_commands/sign.py b/scripts/west_commands/sign.py
index 260adf7db8..c59469bda1 100644
--- a/scripts/west_commands/sign.py
+++ b/scripts/west_commands/sign.py
@@ -430,7 +430,9 @@ class RimageSigner(Signer):
         kernel_name = build_conf.get('CONFIG_KERNEL_BIN_NAME', 'zephyr')
 
         if not target:
-            log.die('rimage target not defined')
+            log.wrn('rimage target not defined')
+            log.wrn('If you want to sign an image, set RIMAGE_TARGET in board.cmake!')
+            return
 
         # TODO: make this a new sign.py --bootloader option.
         if target in ('imx8', 'imx8m'):
```
@dluke62 please, let me know if this works for you.

@marc-hb the RIMAGE_TARGET should be made option - so, no fatal error in case is not defined.
What do you think of the above fix, should I create a PR, or you have a better idea?

dluke62 commented:
> > hi @marc-hb, thank you for your comments. I am getting this build error if I don't disable rimage step in CMakeLists.txt:
> > FAILED: zephyr/zephyr.ri zephyr/build/zephyr/zephyr.ri cd zephyr/build/zephyr/soc/xtensa/nxp_adsp && west sign --if-tool-available --tool rimage --build-dir zephyr/build FATAL ERROR: rimage target not defined ninja: build stopped: subcommand failed.
> > I don't have rimage on my machine, it's not in PATH. Build for imx8m succeeds, I'm getting the warning that rimage is not found.
> 
> Yes, that's because of this: https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/extensions.cmake#L832C10-L832C34, which is defined in each `board.cmake`:
> 
> ```
> ~/data/zephyrproject/zephyr$ grep -rnw board_set_rimage_target 
> cmake/modules/extensions.cmake:832:function(board_set_rimage_target target)
> boards/xtensa/nxp_adsp_imx8x/board.cmake:6:board_set_rimage_target(imx8)
> boards/xtensa/intel_adsp_ace20_lnl/board.cmake:5:board_set_rimage_target(lnl)
> boards/xtensa/intel_adsp_ace15_mtpm/board.cmake:5:board_set_rimage_target(mtl)
> boards/xtensa/intel_adsp_cavs25/board.cmake:13:board_set_rimage_target(tgl)
> boards/xtensa/intel_adsp_cavs25/board.cmake:17:board_set_rimage_target(tgl-h)
> boards/xtensa/nxp_adsp_imx8m/board.cmake:6:board_set_rimage_target(imx8m)
> boards/xtensa/nxp_adsp_imx8/board.cmake:6:board_set_rimage_target(imx8)
> ```
> 
> Check this fix:
> 
> ```
> diff --git a/scripts/west_commands/sign.py b/scripts/west_commands/sign.py
> index 260adf7db8..c59469bda1 100644
> --- a/scripts/west_commands/sign.py
> +++ b/scripts/west_commands/sign.py
> @@ -430,7 +430,9 @@ class RimageSigner(Signer):
>          kernel_name = build_conf.get('CONFIG_KERNEL_BIN_NAME', 'zephyr')
>  
>          if not target:
> -            log.die('rimage target not defined')
> +            log.wrn('rimage target not defined')
> +            log.wrn('If you want to sign an image, set RIMAGE_TARGET in board.cmake!')
> +            return
>  
>          # TODO: make this a new sign.py --bootloader option.
>          if target in ('imx8', 'imx8m'):
> ```
> 
> @dluke62 please, let me know if this works for you.
> 
> @marc-hb the RIMAGE_TARGET should be made option - so, no fatal error in case is not defined. What do you think of the above fix, should I create a PR, or you have a better idea?

Thank you @iuliana-prodan , with your patch the build failure goes away. I am getting the warning message and the build succeeds.

iuliana-prodan commented:
> Thanks @iuliana-prodan , I like your idea. Can you please flesh it a bit more and submit a separate PR? I mean something like this below.
> 
> BTW rimage "target" was not a great name; especially confusing in CMake!
> 

Added https://github.com/zephyrproject-rtos/zephyr/pull/61749

Should I add a second commit to this PR to s/target/board_name?

dluke62 commented:
> Added #61749
> 
> Should I add a second commit to this PR to s/target/board_name?

Thank you @iuliana-prodan ! I've removed first commit that added config option to enable signing. You can add your commit.

marc-hb commented:
> You can add your commit.

Only @dluke62 can push to https://github.com/dluke62/zephyr/tree/dluke62/nxp_rt595_adsp_board :-)

The simplest by far is to wait until #61749 is merged. Then you can perform an empty commit with `git commit -C HEAD --amend` (changes only the date; enough to change the git SHA1) and then force push dluke62/nxp_rt595_adsp_board. CI will merge as usual

dluke62 commented:
> > You can add your commit.
> 
> Only @dluke62 can push to https://github.com/dluke62/zephyr/tree/dluke62/nxp_rt595_adsp_board :-)
> 
> The simplest by far is to wait until #61749 is merged. Then you can perform an empty commit with `git commit -C HEAD --amend` (changes only the date; enough to change the git SHA1) and then force push dluke62/nxp_rt595_adsp_board. Github/CI will merge in pull/61356/merge as usual.

thank you @marc-hb ! will wait until the other pull request is merged.

iuliana-prodan commented:
> > Thanks @iuliana-prodan , I like your idea. Can you please flesh it a bit more and submit a separate PR? I mean something like this below.
> > BTW rimage "target" was not a great name; especially confusing in CMake!
> 
> Added #61749
> 
> Should I add a second commit to this PR to s/target/board_name?

@dluke62 @marc-hb sorry for the confusion - I don't want to push any commit to your PR - meaning #61356.
I was referring to add a new commit to my PR (#61749) to change "target" with "board_name" since this is confusing as mentioned [here](https://github.com/zephyrproject-rtos/zephyr/pull/61356#issuecomment-1687788289) by Marc.

mmahadevan108 commented on boards/arm/mimxrt595_evk/doc/index.rst:
```diff
@@ -174,6 +174,30 @@ Serial Port
+You can build a Zephyr application for the RT500 DSP core using nxp_adsp_rt595
+board. Xtensa toolchain and DSP build configuration are required for this.
+For instructions on how to download the toolchain and debug DSP firmware,
+please refer to this document: `Getting Started with Xplorer for EVK-MIMXRT595`_
```
This should be updated now that the toolchain has been added to the SDK.

dluke62 replied:
updated, thank you!

nashif commented:
This board is now trying to build with `xtensa-nxp_rt5xx_zephyr-elf/bin/xtensa-nxp_rt5xx_zephyr-elf-gcc` which does not exist in the SDK (https://github.com/zephyrproject-rtos/sdk-ng/pull/696 added xtensa-nxp_rt500_adsp_zephyr-elf and is trying to test with nxp_adsp_rt500). We need to fix this.

nashif reviewed (dismissed):
need clarification on board name and toolchain used and sync with the changes required in the sdk.

dbaluta commented:
@nashif any clue where we could find the toolchain built for xtensa-nxp_rt500_adsp_zephyr-elf so that we can download and test with it locally.

@dluke62 PR adding the toolchain was merged. I think you need to update a little bit the patches to make use of the new toolchain.

I think we can consider the SOC family to be rt500 and build on top of that.

dbaluta commented:
@dluke62 toolchain is generated and the new SDK is here:

https://github.com/zephyrproject-rtos/sdk-ng/actions/runs/6002800994?pr=694#artifacts

nashif commented:
or this https://github.com/zephyrproject-rtos/sdk-ng/actions/runs/6001175655#artifacts

dluke62 commented:
> @nashif any clue where we could find the toolchain built for xtensa-nxp_rt500_adsp_zephyr-elf so that we can download and test with it locally.
> 
> @dluke62 PR adding the toolchain was merged. I think you need to update a little bit the patches to make use of the new toolchain.
> 
> I think we can consider the SOC family to be rt500 and build on top of that.

hi @dbaluta , thank you for working on the toolchain. I've updated toolchain name in Kconfig. I am getting a build error when I build with Zephyr toolchain - this include file can't be found <xtensa/config/core-matmap.h>. Looks like for other ADSP targets this file is under modules/hal/xtensa/zephyr/soc/<soc_name>/xtensa/config. There is a number of header files in that folder. Is this something that was generated along with the toolchain?

dluke62 commented:
> This board is now trying to build with `xtensa-nxp_rt5xx_zephyr-elf/bin/xtensa-nxp_rt5xx_zephyr-elf-gcc` which does not exist in the SDK ([zephyrproject-rtos/sdk-ng#696](https://github.com/zephyrproject-rtos/sdk-ng/pull/696) added xtensa-nxp_rt500_adsp_zephyr-elf and is trying to test with nxp_adsp_rt500). We need to fix this.

Thank you @nashif , I've update the toolchain name. My understanding is that the board name should be nxp_adsp_rt595, since there might be more boards based on rt500 SoC. @dbaluta is that right? Or shall I change board name to nxp_adsp_rt500?

dbaluta commented:
> Thank you @nashif , I've update the toolchain name. My understanding is that the board name should be nxp_adsp_rt595, since there might be more boards based on rt500 SoC. @dbaluta is that right? Or shall I change board name to nxp_adsp_rt500?

You are right,  board name should be nxp_adsp_rt595

dluke62 commented:
> 



> > Thank you @nashif , I've update the toolchain name. My understanding is that the board name should be nxp_adsp_rt595, since there might be more boards based on rt500 SoC. @dbaluta is that right? Or shall I change board name to nxp_adsp_rt500?
> 
> You are right, board name should be nxp_adsp_rt595

Thank you for checking @dbaluta! Could you please update the board name in the toolchain? Also, it seems that some of the header files that are needed for build (e.g. xtensa/config/core-matmap.h) are not part of the toolchain PR. Do you know if these files were generated along with the toolchain?

iuliana-prodan commented:
@dluke62 @dbaluta @DerekSnell @dleach02 @nashif based on Derek's proposal [here](https://github.com/zephyrproject-rtos/zephyr/pull/61356#pullrequestreview-1573445368) I believe we should name the toolchain as xtensa-nxp_rt5xx_adsp_zephyr-elf (as the name of the soc series) and keep the board name nxp_adsp_rt595. 
Will this work?

Also, do we need an xtensa overlay in hal_xtensa?
See - https://github.com/zephyrproject-rtos/hal_xtensa/commit/0e577021bb66e644afd067cd9f7c71ab11b62b3d

dbaluta commented:
@dluke62 @iuliana-prodan what's wrong with current toolchain name? It should not matter for building any app.

iuliana-prodan commented:
> @dluke62 @iuliana-prodan what's wrong with current toolchain name? Does it have any impact on the way you write the support for the core?
> 
https://github.com/zephyrproject-rtos/zephyr/pull/61356#issuecomment-1695801395

dbaluta commented:
@dluke62 @DerekSnell @iuliana-prodan I will take care next week to sort things out.

Updating xtensa_hal with missing files and restarting compilation I get an error:

/home/nxf25322/zephyrproject-rtos/modules/hal/xtensa/src/hal/state_asm.S:154: Error: unknown opcode or format name 'addi.a'

This is very unfortunate, and I think because the toolchain we generated is not compatible with Xtensa HAL maintained by the community. 

I think our toolchain is `too new` from the HAL supported by xtensa HAL.

Will sort this out next week.

dbaluta commented:
@DerekSnell @dluke62 @iuliana-prodan this was a wild ride! But I finally managed to compile the firmware binary for nxp_adsp_rt595.

@dluke62 can you please give it a try? You can manually add this PR: https://github.com/zephyrproject-rtos/hal_xtensa/pull/23 into your `zephyrproject-rtos/modules/hal/xtensa` repo.

dluke62 commented:
> @DerekSnell @dluke62 @iuliana-prodan this was a wild ride! But I finally managed to compile the firmware binary for nxp_adsp_rt595.
> 
> @dluke62 can you please give it a try? You can manually add this PR: [zephyrproject-rtos/hal_xtensa#23](https://github.com/zephyrproject-rtos/hal_xtensa/pull/23) into your `zephyrproject-rtos/modules/hal/xtensa` repo.

hi @dbaluta , thank you very much for working on the toolchain! I was able to build and run a sample app on F1 DSP using the header files from your PR. I had to make minor changes to the board files to make it work with PICOLIBC instead of MINIMAL_LIBC. For some reason the toolchain uses PICOLIBC by default instead of MINIMAL_LIBC.  I've updated my PR with the necessary changes.

dbaluta commented:
> hi @dbaluta , thank you very much for working on the toolchain! I was able to build and run a sample app on F1 DSP using the header files from your PR. I had to make minor changes to the board files to make it work with PICOLIBC instead of MINIMAL_LIBC. For some reason the toolchain uses PICOLIBC by default instead of MINIMAL_LIBC. I've updated my PR with the necessary changes.

That's great news! For me it work without any additional changes (but maybe, i had some modifications which I was not aware of).

Anyway, glad that it works! Will let you know when https://github.com/zephyrproject-rtos/hal_xtensa/pull/23 is merged and west.yaml updated.

Regarding picolibc it seems that toolchain common config: https://github.com/zephyrproject-rtos/sdk-ng/blob/main/configs/common.config#L21

enables:
CT_NEWLIB_SRC_CUSTOM=y
CT_COMP_LIBS_NEWLIB_NANO=y
and
CT_COMP_LIBS_PICOLIBC=y

dluke62 commented:
hi @dbaluta , wanted to double check if this PR needs to be updated with a different board name - nxp_adsp_rt595 instead of nxp_adsp_rt500 (https://github.com/zephyrproject-rtos/sdk-ng/pull/696). Looks like CI is using incorrect board name.

dbaluta commented:
@dluke62 I think the board name is OK and it should be like that. Our board is actually an rt595 right? So we should keep the board name `nxp_adsp_rt595`.

The problem I think the CI tries to hint us is that it cannot find the `nxp_adsp_rt595` board because I think we are missing the following file in the current PR.

Can you please add this:

```
--- /dev/null
+++ b/boards/xtensa/nxp_adsp_rt595/nxp_adsp_rt595.yaml
@@ -0,0 +1,9 @@
+identifier: nxp_adsp_rt595
+name: i.MXRT595 DSP
+type: mcu
+arch: xtensa
+toolchain:
+  - zephyr
+testing:
+  only_tags:
+    - kernel
```

dluke62 commented:
> @dluke62 I think the board name is OK and it should be like that. Our board is actually an rt595 right? So we should keep the board name `nxp_adsp_rt595`.
> 
> The problem I think the CI tries to hint us is that it cannot find the `nxp_adsp_rt595` board because I think we are missing something in this PR.
> 
> From: https://github.com/zephyrproject-rtos/zephyr/actions/runs/6229824233/job/16908936405?pr=61356
> 
> ```
> INFO: scripts/twister -c -p nxp_adsp_rt595 --save-tests _test_plan_partial.json
> INFO    - Using Ninja..
> INFO    - Zephyr version: zephyr-v3.4.0-3929-g58e7617072ac
> INFO    - Using 'zephyr' toolchain.
> ERROR   - platform_filter - unrecognized platform - nxp_adsp_rt595
> ```
> 
> Can you please add this:
> 
> ```
> --- /dev/null
> +++ b/boards/xtensa/nxp_adsp_rt595/nxp_adsp_rt595.yaml
> @@ -0,0 +1,9 @@
> +identifier: nxp_adsp_rt595
> +name: i.MXRT595 DSP
> +type: mcu
> +arch: xtensa
> +toolchain:
> +  - zephyr
> +testing:
> +  only_tags:
> +    - kernel
> ```

Thank you @dbaluta ! I've added the yaml file, looks like it worked - 'generate test plan' step is no longer failing.

DerekSnell commented:
> Any idea how to switch to the new zephyr SDK?
> Twister is using zephyr-sdk-0.16.1
> But we need [v0.16.3](https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.16.3)

Adding @hakehuang , our Twister Wizard

marc-hb commented:
Disclaimer: I haven't looked at the logs but last time I checked twister didn't "look" for any particular SDK, all that logic is in CMake (and you probably don't want it in two different places!)

You probably want to look here: https://github.com/zephyrproject-rtos/docker-image

iuliana-prodan commented:
> Disclaimer: I haven't looked at the logs but last time I checked twister didn't "look" for any particular SDK, all that logic is in CMake (and you probably don't want it in two different places!)
> 
> You probably want to look here: https://github.com/zephyrproject-rtos/docker-image

@dluke62 @hakehuang  the docker image seems to use 0.16.3 now - see https://github.com/zephyrproject-rtos/docker-image/blob/main/Dockerfile.ci#L6
@dluke62  can you make a rebase on this PR to trigger the CI?

hakehuang commented:
> Adding @hakehuang , our Twister Wizard

I need upgrade toolchain for this. will feedback once 

> @dluke62 @hakehuang the docker image seems to use 0.16.3 now - see https://github.com/zephyrproject-rtos/docker-image/blob/main/Dockerfile.ci#L6
> @dluke62 can you make a rebase on this PR to trigger the CI?

@iuliana-prodan , just want to confirm, which compiler are we using? see 0.16.3 sdk release, there is not such compiler for xtensa-nxp_rt500_adsp_zephyr-elf
https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.16.3
![image](https://github.com/zephyrproject-rtos/zephyr/assets/800198/79cfa88e-0eb1-411a-adfc-00c998a81e4a)

hakehuang commented:
@iuliana-prodan I update to sdk0.16.3 and see below error in hello world

```
-- Found assembler: /home/ubuntu/zephyr-sdk/zephyr-sdk-0.16.3/xtensa-nxp_rt500_adsp_zephyr-elf/bin/xtensa-nxp_rt500_adsp_zephyr-elf-gcc
-- Configuring done
-- Generating done
-- Build files have been written to: /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/hello_world/build
-- west build: building application
[1/115] Preparing syscall dependency handling

[4/115] Generating include/generated/version.h
-- Zephyr version: 3.4.99 (/home/shared/disk/zephyr_project/zephyr_test/zephyr), build: zephyr-v3.4.0-3989-g7e8e1dff535a
[10/115] Building C object zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj
FAILED: zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj
ccache /home/ubuntu/zephyr-sdk/zephyr-sdk-0.16.3/xtensa-nxp_rt500_adsp_zephyr-elf/bin/xtensa-nxp_rt500_adsp_zephyr-elf-gcc -DCPU_MIMXRT595SFFOC_dsp -DKERNEL -DPICOLIBC_INTEGER_PRINTF_SCANF -D_FORTIFY_SOURCE=1 -D_POSIX_C_SOURCE=200809 -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR__=1 -I/home/shared/disk/zephyr_project/zephyr_test/zephyr/kernel/include -I/home/shared/disk/zephyr_project/zephyr_test/zephyr/arch/xtensa/include -I/home/shared/disk/zephyr_project/zephyr_test/zephyr/include -I/home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/hello_world/build/zephyr/include/generated -I/home/shared/disk/zephyr_project/zephyr_test/zephyr/soc/xtensa/nxp_adsp/rt5xx -I/home/shared/disk/zephyr_project/zephyr_test/zephyr/soc/xtensa/nxp_adsp/rt5xx/include -I/home/shared/disk/zephyr_project/zephyr_test/zephyr/soc/xtensa/nxp_adsp/common/include -I/home/shared/disk/zephyr_project/zephyr_test/modules/hal/xtensa/include -fno-strict-aliasing -Os -imacros /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/hello_world/build/zephyr/include/generated/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always --sysroot=/home/ubuntu/zephyr-sdk/zephyr-sdk-0.16.3/xtensa-nxp_rt500_adsp_zephyr-elf/xtensa-nxp_rt500_adsp_zephyr-elf -imacros /home/shared/disk/zephyr_project/zephyr_test/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/hello_world=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/shared/disk/zephyr_project/zephyr_test/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/shared/disk/zephyr_project/zephyr_test=WEST_TOPDIR -ffunction-sections -fdata-sections -mlongcalls --specs=picolibc.specs -std=c99 -MD -MT zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj -MF zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj.d -o zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj -c /home/shared/disk/zephyr_project/zephyr_test/zephyr/arch/xtensa/core/offsets/offsets.c
In file included from /home/shared/disk/zephyr_project/zephyr_test/zephyr/include/zephyr/arch/xtensa/arch.h:28,
                 from /home/shared/disk/zephyr_project/zephyr_test/zephyr/include/zephyr/arch/cpu.h:27,
                 from /home/shared/disk/zephyr_project/zephyr_test/zephyr/include/zephyr/kernel_includes.h:37,
                 from /home/shared/disk/zephyr_project/zephyr_test/zephyr/include/zephyr/kernel.h:17,
                 from /home/shared/disk/zephyr_project/zephyr_test/zephyr/include/zephyr/pm/device.h:11,
                 from /home/shared/disk/zephyr_project/zephyr_test/zephyr/kernel/include/kernel_offsets.h:7,
                 from /home/shared/disk/zephyr_project/zephyr_test/zephyr/arch/xtensa/core/offsets/offsets.c:7:
/home/shared/disk/zephyr_project/zephyr_test/modules/hal/xtensa/include/xtensa/config/core.h:51:10: fatal error: xtensa/config/core-matmap.h: No such file or directory
   51 | #include <xtensa/config/core-matmap.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /home/ubuntu/.local/bin/cmake --build /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/hello_world/build


```

dbaluta commented:
@hakehuang This should have been fixed with:

https://github.com/zephyrproject-rtos/hal_xtensa/pull/23

Can you please check that Zephyr west.yml with your repo pointa to updated version of xtensa hal?

hakehuang commented:
> Can you please check that Zephyr west.yml with your repo points to updated version of xtensa hal?

hi @dbaluta , the tipical practices is you need to modify your PR with west.yml point to the hal pull request, and mark this one as DNM, till the HAL PR is merged. it is very difficult to track the HAL and zephyr code, if you do it separately.

iuliana-prodan commented:
@dluke62 can you please add a new commit to this PR to update the `hal_xtensa` in `west.yml`, like this:
```
diff --git a/west.yml b/west.yml
index 936b90ff2e..a28839d721 100644
--- a/west.yml
+++ b/west.yml
@@ -255,7 +255,7 @@ manifest:
       groups:
         - hal
     - name: hal_xtensa
-      revision: 86b7ddf984d54f34b79f52b328a4fecdae307956
+      revision: e6da34fc07dfe96161ab8743f5dbeb6e6307ab93
       path: modules/hal/xtensa
       groups:
         - hal
```

The CI should pass after this.
cc: @hakehuang

dbaluta commented:
@iuliana-prodan I think it is better if you create a separate PR to update the xtensa hal then @dluke62 will rebase his changes.

iuliana-prodan commented:
> e6da34fc07dfe96161ab8743f5dbeb6e6307ab93

@dluke62 @dbaluta I've created https://github.com/zephyrproject-rtos/zephyr/pull/63095

iuliana-prodan commented:
> > @iuliana-prodan I think it is better if you create a separate PR to update the xtensa hal then @dluke62 will rebase his changes.
> 
@dluke62 @dbaluta I've created #63095

dbaluta commented:
@hakehuang can you please try again.

https://github.com/zephyrproject-rtos/zephyr/pull/6309

is mergef.

hakehuang commented:
> @hakehuang can you please try again.
> 
> #63095
> 
> is merged.

@dbaluta , you still need update the west.yml to direct to the right hal_nxp commit.

iuliana-prodan commented:
> > @hakehuang can you please try again.
> > #63095
> > is merged.
> 
> @dbaluta , you still need update the west.yml to direct to the right hal_nxp commit.

@hakehuang the hal_nxp from west.yml has all the needed support for this PR. So, no changes there.

Can you please try to run CI on this PR?

@dluke62 can you rebase or amend this PR in order to trigger the Zephyr CI.

hakehuang commented:
> @hakehuang the hal_nxp from west.yml has all the needed support for this PR. So, no changes there.
> Can you please try to run CI on this PR?

@iuliana-prodan in this case, you need rebase this PR, I try to rebase for you, but I just not get permission

marc-hb commented:
> I try to rebase for you, but I just not get permission.

Indeed you cannot change someone else's clone.

The way to solve this problem (for instance: when someone is out of office), is to fetch their branch (`git fetch origin pull/61356`) and submit a new PR. This preserves their name.

iuliana-prodan commented:
> > I try to rebase for you, but I just not get permission.
> 
> Indeed you cannot change someone else's clone.
> 
> The way to solve this problem (for instance: when someone is out of office), is to fetch their branch (`git fetch origin pull/61356`) and submit a new PR. This preserves their name.

@hakehuang this is not my PR.
Let's wait a bit for @dluke62, maybe he can update this (/his) PR to trigger the CI.
All I can do is what @marc-hb suggested above, but I would like to wait for @dluke62 :)
I don't want to duplicate this PR.

dluke62 commented:
> > > I try to rebase for you, but I just not get permission.
> > 
> > 
> > Indeed you cannot change someone else's clone.
> > The way to solve this problem (for instance: when someone is out of office), is to fetch their branch (`git fetch origin pull/61356`) and submit a new PR. This preserves their name.
> 
> @hakehuang this is not my PR. Let's wait a bit for @dluke62, maybe he can update this (/his) PR to trigger the CI. All I can do is what @marc-hb suggested above, but I would like to wait for @dluke62 :) I don't want to duplicate this PR.

thank you guys! I am rebasing the PR, will update it soon.

hakehuang commented:
a build warning:

```
-- Including generated dts.cmake file: /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/hello_world/build/zephyr/dts.cmake
/home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/hello_world/build/zephyr/zephyr.dts:18.8-26.5: Warning (simple_bus_reg): /soc/cpus: missing or empty reg/ranges property
Parsing /home/shared/disk/zephyr_project/zephyr_test/zephyr/Kconfig

```

LaurentiuM1234 commented on soc/xtensa/nxp_adsp/rt5xx/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,46 @@
+config SYS_CLOCK_TICKS_PER_SEC
+	default 1000
+
+config DYNAMIC_INTERRUPTS
```
Isn't this disabled by default?

LaurentiuM1234 commented on soc/xtensa/nxp_adsp/rt5xx/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,46 @@
+config DCACHE
+	default n
+
+config CACHE_MANAGEMENT
```
Same here.

LaurentiuM1234 commented on soc/xtensa/nxp_adsp/rt5xx/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,46 @@
+config CACHE_MANAGEMENT
+	default n
+
+config LOG
```
Same here.

LaurentiuM1234 commented on soc/xtensa/nxp_adsp/Kconfig.defconfig:
```diff
@@ -1,9 +1,29 @@
-	def_bool y
+	default y
+
+config SMP
```
This also seems to be disabled by default. Could probably remove this altogether from all xtensa-based NXP SoCs.

LaurentiuM1234 commented on soc/xtensa/nxp_adsp/Kconfig.defconfig:
```diff
@@ -1,9 +1,29 @@
+config KERNEL_ENTRY
+	default "__start"
+
+config MULTI_LEVEL_INTERRUPTS
```
Same here.

LaurentiuM1234 commented on soc/xtensa/nxp_adsp/Kconfig.defconfig:
```diff
@@ -1,9 +1,29 @@
+config MULTI_LEVEL_INTERRUPTS
+	default n
+
+config 2ND_LEVEL_INTERRUPTS
```
Same here.

LaurentiuM1234 commented on boards/xtensa/nxp_adsp_rt595/nxp_adsp_rt595_defconfig:
```diff
@@ -0,0 +1,7 @@
+
+CONFIG_GEN_ISR_TABLES=y
+CONFIG_GEN_IRQ_VECTOR_TABLE=n
+CONFIG_XTENSA_SMALL_VECTOR_TABLE_ENTRY=y
```
Aren't all of these configs SoC quirks? If so we could probably move them to `soc/xtensa/nxp_adsp/Kconfig.defconfig` as it seems like all xtensa-based NXP SoCs use them.

LaurentiuM1234 reviewed (commented):
A few nitpicks/suggestions

dbaluta commented:
@nashif I think we are ready to merge this. @LaurentiuM1234 we can fix the nitpicks in the future PRs.

dluke62 commented:
> a build warning:
> 
> ```
> -- Including generated dts.cmake file: /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/hello_world/build/zephyr/dts.cmake
> /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/hello_world/build/zephyr/zephyr.dts:18.8-26.5: Warning (simple_bus_reg): /soc/cpus: missing or empty reg/ranges property
> Parsing /home/shared/disk/zephyr_project/zephyr_test/zephyr/Kconfig
> ```
> 
> others are ok
> 
> ```
> scripts/twister -p nxp_adsp_rt595 -T tests/kernel/ --build-only
> Renaming output directory to /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out.1
> INFO    - Using Ninja..
> INFO    - Zephyr version: zephyr-v3.4.0-4456-ge936ecbb074d
> INFO    - Using 'zephyr' toolchain.
> INFO    - Building initial testsuite list...
> INFO    - Writing JSON report /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/testplan.json
> INFO    - JOBS: 16
> INFO    - Adding tasks to the queue...
> INFO    - Added initial list of jobs to queue
> INFO    - Total complete:  177/ 177  100%  skipped:   90, failed:    0, error:    0
> INFO    - 177 test scenarios (177 test instances) selected, 90 configurations skipped (59 by static filter, 31 at runtime).
> INFO    - 87 of 177 test configurations passed (100.00%), 0 failed, 0 errored, 90 skipped with 0 warnings in 380.11 seconds
> INFO    - 0 test configurations executed on platforms, 87 test configurations were only built.
> INFO    - Saving reports...
> INFO    - Writing JSON report /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/twister.json
> INFO    - Writing xunit report /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/twister.xml...
> INFO    - Writing xunit report /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/twister_report.xml...
> INFO    - Run completed
> ```

Thank you, fixed the warning.

dluke62 commented:
@nashif would you mind taking another look? CI is now green.

dluke62 commented:
Hi @dbaluta! Do you know if there is a way to move forward with this PR? Are we blocked by Anas's change request?

dbaluta commented:
@dluke62 I talked with @nashif  on discord and he will pick this up once the 3.5 release is done. Will ping him again.

dluke62 commented:
> @dluke62 I talked with @nashif on discord and he will pick this up once the 3.5 release is out. Will ping him again.

thank you!!
